security/advancedtls/advancedtls.go

```diff
@@ -251,85 +284,75 @@ func (c *advancedTLSCreds) OverrideServerName(serverNameOverride string) error {
+// 1. does not have a good support on root cert reloading.
+// 2. will ignore basic certificate check when setting InsecureSkipVerify to true.
 func buildVerifyFunc(c *advancedTLSCreds,
 	serverName string,
 	rawConn net.Conn) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
 	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-		// If users didn't specify either rootCAs or getRootCAs on client side,
-		// as we see some use cases such as https://github.com/grpc/grpc/pull/20530,
-		// instead of failing, we just don't validate the server cert and let
-		// application decide via VerifyPeer
-		if c.isClient && c.config.RootCAs == nil && c.getRootCAs == nil {
-			if c.verifyFunc != nil {
-				_, err := c.verifyFunc(&VerificationFuncParams{
-					ServerName:     serverName,
-					RawCerts:       rawCerts,
-					VerifiedChains: verifiedChains,
+		chains := verifiedChains
+		if c.vType == CertAndHostVerification || c.vType == CertVerification {
+			// perform possible trust credential reloading and certificate check
+			var rootCAs *x509.CertPool
+			if c.isClient {
+				rootCAs = c.config.RootCAs
+			} else {
+				rootCAs = c.config.ClientCAs
+			}
+			// Reload root CA certs.
+			if rootCAs == nil && c.getRootCAs != nil {
+				results, err := c.getRootCAs(&GetRootCAsParams{
+					RawConn:  rawConn,
+					RawCerts: rawCerts,
 				})
-				return err
+				if err != nil {
+					return err
+				}
+				rootCAs = results.TrustCerts
 			}
-		}
-		var rootCAs *x509.CertPool
-		if c.isClient {
-			rootCAs = c.config.RootCAs
-		} else {
-			rootCAs = c.config.ClientCAs
-		}
-		// reload root CA certs
-		if rootCAs == nil && c.getRootCAs != nil {
-			results, err := c.getRootCAs(&GetRootCAsParams{
-				RawConn:  rawConn,
-				RawCerts: rawCerts,
-			})
-			if err != nil {
-				return err
+			// Verify peers' certificates against RootCAs and get verifiedChains.
+			certs := make([]*x509.Certificate, len(rawCerts))
+			for i, asn1Data := range rawCerts {
+				cert, err := x509.ParseCertificate(asn1Data)
+				if err != nil {
+					return err
+				}
+				certs[i] = cert
 			}
-			rootCAs = results.TrustCerts
-		}
-		// verify peers' certificates against RootCAs and get verifiedChains
-		certs := make([]*x509.Certificate, len(rawCerts))
-		for i, asn1Data := range rawCerts {
-			cert, err := x509.ParseCertificate(asn1Data)
+			opts := x509.VerifyOptions{
+				Roots:         rootCAs,
+				CurrentTime:   time.Now(),
+				Intermediates: x509.NewCertPool(),
+			}
+			if !c.isClient {
+				opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
+			} else {
+				opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+			}
+			for _, cert := range certs[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			// Perform default hostname check if specified.
+			var err error
+			if c.isClient && c.vType == CertAndHostVerification && serverName != "" {
+				opts.DNSName = serverName
+			}
```

Comment:
I think it would make the code a little bit more readable if you could construct things like the intermediate cert pool, the key usages, dnsName etc and then in one swoop construct your x509.VerifyOptions